File: tools/src/main/java/org/spine3/tools/javadoc/ExcludeInternalDoclet.java
/*
 * Copyright 2017, TeamDev Ltd. All rights reserved.
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.spine3.tools.javadoc;

import com.sun.javadoc.Doc;
import com.sun.javadoc.RootDoc;
import com.sun.tools.doclets.standard.Standard;
import com.sun.tools.javadoc.Main;
import org.spine3.Internal;

import java.lang.reflect.Array;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.util.ArrayList;
import java.util.List;

/**
 * Extension of {@linkplain Standard} doclet, which excludes {@linkplain Internal}-annotated components.
 * Based on <a href="http://sixlegs.com/blog/java/exclude-javadoc-tag.html">this</a> topic.
 *
 * @author Dmytro Grankin
 */
@SuppressWarnings("ExtendsUtilityClass")
public class ExcludeInternalDoclet extends Standard {

    private final ExcludePrinciple excludePrinciple;

    private ExcludeInternalDoclet(ExcludePrinciple excludePrinciple) {
        super();
        this.excludePrinciple = excludePrinciple;
    }

    /**
     * Entry point for the Javadoc tool.
     *
     * @param args the command-line parameters
     */
    public static void main(String[] args) {
        final String name = ExcludeInternalDoclet.class.getName();
        Main.execute(name, name, args);
    }

    /**
     * The "start" method as required by Javadoc.
     *
     * @param root the root of the documentation tree.
     * @return {@code true} if the doclet ran without encountering any errors.
     */
    @SuppressWarnings("unused") // called by com.sun.tools.javadoc.Main
    public static boolean start(RootDoc root) {
        final ExcludePrinciple excludePrinciple = new ExcludeInternalPrinciple(root);
        final ExcludeInternalDoclet doclet = new ExcludeInternalDoclet(excludePrinciple);
        return Standard.start((RootDoc) doclet.process(root, RootDoc.class));
    }

    private Object process(Object obj, Class expect) {
        if (obj == null) {
            return null;
        }

        if (obj.getClass().getName().startsWith("com.sun.")) {
            final Class cls = obj.getClass();
            return Proxy.newProxyInstance(cls.getClassLoader(), cls.getInterfaces(), new ExcludeHandler(obj));
        } else if (obj instanceof Object[] && expect.getComponentType() != null) {
            final Class componentType = expect.getComponentType();
            final Object[] array = (Object[]) obj;
            final List<Object> list = new ArrayList<>();
            for (Object entry : array) {
                if (!(entry instanceof Doc && excludePrinciple.exclude((Doc) entry))) {
                    list.add(process(entry, componentType));
                }
            }
            return list.toArray((Object[]) Array.newInstance(componentType, list.size()));
        } else {
            return obj;
        }
    }

    private class ExcludeHandler implements InvocationHandler {

        private final Object target;

        private ExcludeHandler(Object target) {
            this.target = target;
        }

        @Override
        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
            if (args != null) {
                String methodName = method.getName();
                if ("compareTo".equals(methodName)
                        || "equals".equals(methodName)
                        || "overrides".equals(methodName)
                        || "subclassOf".equals(methodName)) {
                    args[0] = unwrap(args[0]);
                }
            }
            try {
                return process(method.invoke(target, args), method.getReturnType());
            } catch (InvocationTargetException e) {
                throw e.getCause();
            }
        }

        private Object unwrap(Object proxy) {
            if (proxy instanceof Proxy) {
                return ((ExcludeHandler) Proxy.getInvocationHandler(proxy)).target;
            }
            return proxy;
        }
    }
}

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javadoc/ExcludeInternalDoclet.kt
/*
 * Copyright 2021, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.internal.gradle.javadoc

import io.spine.internal.gradle.javadoc.ExcludeInternalDoclet.Companion.taskName
import io.spine.internal.gradle.sourceSets
import org.gradle.api.Project
import org.gradle.api.artifacts.Configuration
import org.gradle.api.tasks.javadoc.Javadoc
import org.gradle.external.javadoc.StandardJavadocDocletOptions

/**
 * The doclet which removes Javadoc for `@Internal` things in the Java code.
 */
class ExcludeInternalDoclet(val version: String) {

    private val dependency = "io.spine.tools:spine-javadoc-filter:${version}"

    companion object {

        /**
         * The name of the custom configuration in scope of which the exclusion of
         * `@Internal` types is performed.
         */
        private const val configurationName = "excludeInternalDoclet"

        /**
         * The fully-qualified class name of the doclet.
         */
        const val className = "io.spine.tools.javadoc.ExcludeInternalDoclet"

        /**
         * The name of the helper task which configures the Javadoc processing
         * to exclude `@Internal` types.
         */
        const val taskName = "noInternalJavadoc"

        private fun createConfiguration(project: Project): Configuration {
            return project.configurations.create(configurationName)
        }
    }

    /**
     * Creates a custom Javadoc task for the [project] which excludes the the types
     * annotated as `@Internal`.
     *
     * The task is registered under [taskName].
     */
    fun registerTaskIn(project: Project) {
        val configuration = addTo(project)
        project.appendCustomJavadocTask(configuration)
    }

    /**
     * Creates a configuration for the doclet in the given project and adds it to its dependencies.
     *
     * @return added configuration
     */
    private fun addTo(project: Project): Configuration {
        val configuration = createConfiguration(project)
        project.dependencies.add(configuration.name, dependency)
        return configuration
    }
}

private fun Project.appendCustomJavadocTask(excludeInternalDoclet: Configuration) {
    val javadocTask = tasks.javadocTask()
    tasks.register(taskName, Javadoc::class.java) {

        source = sourceSets.getByName("main").allJava.filter {
            !it.absolutePath.contains("generated")
        }.asFileTree

        classpath = javadocTask.classpath

        options {
            encoding = JavadocConfig.encoding.name

            // Doclet fully qualified name.
            doclet = ExcludeInternalDoclet.className

            // Path to the JAR containing the doclet.
            docletpath = excludeInternalDoclet.files.toList()
        }

        val docletOptions = options as StandardJavadocDocletOptions
        JavadocConfig.registerCustomTags(docletOptions)
    }
}
